### PR TITLE
[SPARK-12542] [SQL] support except/intersect in HiveQl

### PR DESCRIPTION
**Pull request by davies**
Parse the SQL query with except/intersect in FROM clause for HivQL.

### CODE REVIEW

**Comment by davies** [comment]
@hvanhovell This patch can't be compiled, failed with:

```
[error] ~spark/sql/hive/target/scala-2.10/src_managed/main/org/apache/spark/sql/parser/SparkSqlParser_IdentifiersParser.java:11056:  ERROR: too long
[error] 	static final String[] DFA5_transitionS = {
[error] 	                      ^
```

**Comment by hvanhovell** [comment]
@davies I haven't seen this one before. We can take a look at the generated source...

**Comment by hvanhovell** [comment]
@davies I just tried this. If you add this rule the identifier parser grows beyond 64K rows which the JVM can't handle. The short term solution is to split the identifier parser (see @viirya's PR).

In the long term we might need to check if some rules aren't blowing things out of proportion.

**Comment by davies** [comment]
@hvanhovell Thanks, I will wait for @viirya's PR go first.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala [hunk 1]
```diff
@@ -195,4 +195,11 @@ class HiveQlSuite extends SparkFunSuite with BeforeAndAfterAll {
 
     assert(plan.children.head.asInstanceOf[Generate].generator.isInstanceOf[JsonTuple])
   }
+
+  test("parse union distinct/except/intersect") {
+    HiveQl.parseSql("select a, b from t1 union all t2 union all t3")
+    HiveQl.parseSql("select a, b from t1 union distinct t2 union distinct t3")
```
Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala]: do we support just union, i.e. alias for union distinct?

Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala]: Hive 1.2's parser should support union.

Comment (rxin) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/HiveQlSuite.scala]: Actually this test case shouldn't be here? It should just go into the normal sql test case now.

**Comment by rxin** [comment]
LGTM. Would be great for @hvanhovell to take a quick look too.

**Comment by davies** [comment]
The following query does not work:
```
(select 1 as a) union (select 2 as a)
```
I had not figure out a way to fix this.

PS: It seems that HiveParse is not a good stand point. lots of corner cases do not work.

**Comment by rxin** [comment]
Let's open a jira ticket to track that and see if some open source contributors can pick it up.

**Comment by davies** [comment]
https://issues.apache.org/jira/browse/SPARK-12687

**Comment by hvanhovell** [comment]
LGTM

**Comment by rxin** [comment]
Merging this in. Thanks!

**Pull request closed without merging**